src/ucs/datastruct/ptr_array.h

```diff
@@ -120,13 +146,169 @@ void *ucs_ptr_array_replace(ucs_ptr_array_t *ptr_array, unsigned index, void *ne
                     (UCS_V_INITIALIZED(_var), 0) : \
                     !__ucs_ptr_array_is_free(_var = (void*)((_ptr_array)->start[_index])))
 
+/**
+ * Check if element is free, and calculate how many free elements are ahead.
+ *
+ * @param [in] _elem      An element in the ptr array.
+ * @param [out] size_free Pointer to the returned number of free elements ahead.
+ * @return                1 if the element is free and 0 if it's occupied.
+ */
+static inline int
+__ucs_ptr_array_is_free_and_get_step_size(ucs_ptr_array_elem_t _elem,
+                                          uint32_t *_size_free)
+{
+    int _is_free = ((uintptr_t)(_elem) & UCS_PTR_ARRAY_FLAG_FREE);
+
+    if (_is_free) {
+       *_size_free = _elem >> UCS_PTR_ARRAY_SIZE_FREE_SHIFT;
+    } else {
+       *_size_free = 1;
+    }
+    ucs_assert(*_size_free > 0);
+
+    return _is_free;
+}
 
 /**
  * Iterate over all valid elements in the array.
+ *
+ * @param [out] _var      Pointer to current array element in the foreach.
+ * @param [out] _index    Index variable to use as iterator (unsigned).
+ * @param [in] ptr_array  Pointer to a ptr array.
```

Comment:
`_ptr_array`
+ one more whitespace after `[in]` to align with parameters above that have `[out]`

Reply:
Fixed.